src/main.rs

```diff
@@ -0,0 +1,148 @@
+            if let (Some(source), Some(destination)) =
+                (yaml_config.source.clone(), yaml_config.destination.clone())
+            {
+                (source, destination)
+            } else {
+                return Err(miette::miette!("Source and target must be specified"));
+            }
+        }
+        _ => unreachable!("prevented by clap"),
+    };
+
+    let subdirs = if let Some(subdirs) = cli_config.subdir {
+        Some(subdirs)
+    } else {
+        yaml_config.subdirs.clone()
+    };
+
+    let mode = match (yaml_config.include, yaml_config.exclude) {
+        (Some(include), Some(exclude)) => MirrorMode::IncludeExclude(include, exclude),
+        (Some(include), None) => MirrorMode::OnlyInclude(include),
+        (None, Some(exclude)) => MirrorMode::AllButExclude(exclude),
+        (None, None) => MirrorMode::All,
+    };
+
+    let s3_config_destination = if let (Some(endpoint_url), Some(region), Some(force_path_style)) = (
+        cli_config.s3_endpoint_url_destination,
+        cli_config.s3_region_destination,
+        cli_config.s3_force_path_style_destination,
+    ) {
+        Some(S3Config {
+            endpoint_url,
+            region,
+            force_path_style,
+        })
+    } else if let Some(s3_config_source_dest) = yaml_config.s3_config.clone() {
+        if let Some(s3_config) = s3_config_source_dest.destination {
+            Some(S3Config {
+                endpoint_url: s3_config.endpoint_url,
+                region: s3_config.region,
+                force_path_style: s3_config.force_path_style,
+            })
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+    let s3_config_source = if let (Some(endpoint_url), Some(region), Some(force_path_style)) = (
+        cli_config.s3_endpoint_url_source,
+        cli_config.s3_region_source,
+        cli_config.s3_force_path_style_source,
+    ) {
+        Some(S3Config {
+            endpoint_url,
+            region,
+            force_path_style,
+        })
+    } else if let Some(s3_config_source_dest) = yaml_config.s3_config {
+        if let Some(s3_config) = s3_config_source_dest.source {
+            Some(S3Config {
+                endpoint_url: s3_config.endpoint_url,
+                region: s3_config.region,
+                force_path_style: s3_config.force_path_style,
+            })
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    let s3_credentials_destination = if let (Some(access_key_id), Some(secret_access_key)) = (
+        cli_config.s3_access_key_id_destination,
+        cli_config.s3_secret_access_key_destination,
+    ) {
+        Some(S3Credentials {
+            access_key_id,
+            secret_access_key,
+            session_token: cli_config.s3_session_token_destination,
+        })
+    } else {
+        None
+    };
+
+    // let s3_credentials_source = if let (Some(access_key_id), Some(secret_access_key)) = (
```

Comment:
remove?

Reply:
this was left in there intentionally. in order to support s3-credentials for the source, i need to inject a custom in-memory authentication backend in rattler which i'm planning to add support for soon.